[corecheck/containers/kubelet] Capture main /metrics endpoint metrics

sblumenthal opened the pull request:
### What does this PR do?

This PR re-implements the metric collection logic from the python kubelet check and adapts it to the new kubelet corecheck

### Motivation

[CONTINT-99](https://datadoghq.atlassian.net/browse/CONTINT-99)

### Additional Notes

[This file](https://github.com/DataDog/integrations-core/blob/master/kubelet/datadog_checks/kubelet/kubelet.py) is the one which has been ported over.

In order to support these metrics, the remaining functionality* from the python prometheus mixin needed to be migrated over. As such, the bulk of the (non-test fixture) additions in this PR are changes to the prometheus provider, and they are simply re-implementing functionality which exist [here](https://github.com/DataDog/integrations-core/blob/master/datadog_checks_base/datadog_checks/base/checks/openmetrics/mixins.py)

### Possible Drawbacks / Trade-offs

Similar to the PR which added support for the probes endpoint, the goal here is not to fully implement the python prometheus code, as many of the codepaths in that code are not valid for the kubelet check. As such, the disclaimer at the top of that file is still accurate: it should not be assumed that this implementation has full 1 to 1 feature parity.

### Describe how to test/QA your changes

Same as https://github.com/DataDog/datadog-agent/pull/18536:

Deploy the agent image using the public helm chart with the following values:
```
datadog:
  confd:
    kubelet_core.yaml: |
      init_config:
      instances:
        - min_collection_interval: 20
```

and validate that some of the metrics which have been added in this PR are being generated, either via the UI or via:
```
agent check -r kubelet_core --table
```

Metrics added are:
```
		"kubernetes_core.apiserver.certificate.expiration.count",
		"kubernetes_core.apiserver.certificate.expiration.sum",
		"kubernetes_core.rest.client.requests",
		"kubernetes_core.rest.client.latency.count",
		"kubernetes_core.rest.client.latency.sum",
		"kubernetes_core.kubelet.runtime.operations",
		"kubernetes_core.kubelet.runtime.errors",
		"kubernetes_core.kubelet.runtime.operations.duration.sum",
		"kubernetes_core.kubelet.runtime.operations.duration.count",
		"kubernetes_core.kubelet.runtime.operations.duration.quantile",
		"kubernetes_core.kubelet.docker.operations",
		"kubernetes_core.kubelet.docker.errors",
		"kubernetes_core.kubelet.docker.operations.duration.sum",
		"kubernetes_core.kubelet.docker.operations.duration.count",
		"kubernetes_core.kubelet.docker.operations.duration.quantile",
		"kubernetes_core.kubelet.network_plugin.latency.sum",
		"kubernetes_core.kubelet.network_plugin.latency.count",
		"kubernetes_core.kubelet.volume.stats.available_bytes",
		"kubernetes_core.kubelet.volume.stats.capacity_bytes",
		"kubernetes_core.kubelet.volume.stats.used_bytes",
		"kubernetes_core.kubelet.volume.stats.inodes",
		"kubernetes_core.kubelet.volume.stats.inodes_free",
		"kubernetes_core.kubelet.volume.stats.inodes_used",
		"kubernetes_core.kubelet.evictions",
		"kubernetes_core.kubelet.pod.start.duration.sum",
		"kubernetes_core.kubelet.pod.start.duration.count",
		"kubernetes_core.kubelet.pod.worker.start.duration.sum",
		"kubernetes_core.kubelet.pod.worker.start.duration.count",
		"kubernetes_core.go_threads",
		"kubernetes_core.go_goroutines",
```

Depending on your environment, not all of them may show up, but some should always, such as `go_threads` and `go_goroutines`

Deploy a statefulset to your cluster, such as `helm install --create-namespace -n redis redis-test oci://registry-1.docker.io/bitnamicharts/redis` and validate that the volume metrics show up with pod tags attached (such as `kube_namespace` and `kube_stateful_set` or others which would not be showing up by default on other metrics)

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.


[CONTINT-99]: https://datadoghq.atlassian.net/browse/CONTINT-99?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

sblumenthal committed:
Add mostly complete kubelet /metrics endpoint implementation

- This commit also fleshes out the prometheus code a bit more to be closer
  in parity to the functionality of the python prometheus code

sblumenthal committed:
Add helper functions to get container ID by pod name

sblumenthal committed:
Implement final kubelet metric transformers

sblumenthal committed:
Enable metrics endpoint in kubelet check

sblumenthal committed:
Fix config parsing, add default config to send buckets, fix metric

sblumenthal committed:
Fix panic when map is nil

sblumenthal committed:
Add test fixture files

- Committing these separately, as I am noticing that some codepaths do not
  have corresponding metrics, and I want the changes that I make to these
  files to be clear and visible

sblumenthal committed:
Fix prometheus metric map parsing, move cache function to common

sblumenthal committed:
Add first pass at tests, add check to make sure metric value is valid

sblumenthal committed:
Add remaining tests to kubelet metrics provider

sblumenthal committed:
Address some linter errors

sblumenthal committed:
Merge remote-tracking branch 'origin' into sblumenthal/cont-3870-implement-main-metrics-endpoint

sblumenthal committed:
Fix remaining linting errors

sblumenthal committed:
Fix failing test after merge

sblumenthal committed:
Remove remaining TODOs

sblumenthal committed:
Fix formatting issue because my IDE has failed me

clamoriniere commented on pkg/collector/corechecks/containers/kubelet/provider/pod/provider.go:
```diff
@@ -70,7 +70,7 @@ func (p *Provider) Provide(kc kubelet.KubeUtilInterface, sender sender.Sender) e
 
 	for _, pod := range pods.Items {
-		//for _, container := range pod.Spec.Containers {
+		common.CachePodTagsByPVC(pod)
```
I'm wondering why we need to push the pod tag in a cache instead of getting them locally in this for loop? 
do we reuse them in another part of the code?

What I see in this `CachePodTagsByPVC` function is that we always set the entry in the cache even it is already in, so I don't know if we improve performance by caching this info.

sblumenthal replied:
They're inserted into the cache here, and fetched from the cache for the volume metrics transformed from the /metrics endpoint.

I opted for a cache to keep the interface simple and to reduce the need to refactor the function signatures of all of the metric providers. But I can also just as easily add this map as a field on the pod provider and the /metrics provider, and it could accomplish the same thing. Using the cache allowed me to also take advantage of the existing expiration timer logic built into that so that I did not need to re-implement something for this.

The cache here is not being used for performance reasons, it is being used to share data, which could be a good reason to not actually use a cache then over a map. As far as always setting the entry, it is for 2 reasons:
1) in case pod tags change
2) to refresh the expiration timer on the cached entry

sblumenthal replied:
I have updated the PR and removed the cache, replacing it instead with a map that is shared throughout the course of the check and is cleared after every check run. Let me know what you think

sblumenthal committed:
Merge remote-tracking branch 'origin' into sblumenthal/cont-3870-implement-main-metrics-endpoint

# Conflicts:
#	pkg/collector/corechecks/containers/kubelet/common/config.go
#	pkg/collector/corechecks/containers/kubelet/kubelet.go
#	pkg/collector/corechecks/containers/kubelet/provider/probe/provider.go
#	pkg/collector/corechecks/containers/kubelet/provider/prometheus/provider.go

sblumenthal committed:
Remove caching layer in favor of shared map

- Removes the use of the cache.Cache pkg as a caching layer for pod PVC tags
  and replaced it with a shared struct that contains a map representing this
  data. This data is cleared after every check run, which mimics the behavior
  of the old python check.

clamoriniere reviewed (approved):
Awesome work 👏
